guides/source/active_record_callbacks.md

```diff
@@ -3,108 +3,137 @@
+Implementation can be done in a multitude of ways like using ordinary methods,
+blocks and procs, or defining custom callback objects using classes or modules.
+Let's go through each of these implementation techniques.
 
-In order to use the available callbacks, you need to register them. You can implement the callbacks as ordinary methods and use a macro-style class method to register them as callbacks:
+You can implement the callbacks as a **macro-style method that calls an ordinary
+method** for registration.
 
 ```ruby
 class User < ApplicationRecord
-  validates :login, :email, presence: true
+  validates :username, :email, presence: true
 
-  before_validation :ensure_login_has_a_value
+  before_validation :ensure_username_has_value
 
   private
-    def ensure_login_has_a_value
-      if login.blank?
-        self.login = email unless email.blank?
+    def ensure_username_has_value
+      if username.blank?
+        self.username = email
       end
     end
 end
 ```
 
-The macro-style class methods can also receive a block. Consider using this style if the code inside your block is so short that it fits in a single line:
+The **macro-style class methods can also receive a block**. Consider using this
+style if the code inside your block is so short that it fits in a single line:
 
 ```ruby
 class User < ApplicationRecord
-  validates :login, :email, presence: true
+  validates :username, :email, presence: true
 
-  before_create do
-    self.name = login.capitalize if name.blank?
+  before_validation do
+    self.username = email if username.blank?
   end
 end
 ```
 
-Alternatively you can pass a proc to the callback to be triggered.
+Alternatively, you can **pass a proc to the callback** to be triggered.
 
 ```ruby
 class User < ApplicationRecord
-  before_create ->(user) { user.name = user.login.capitalize if user.name.blank? }
+  validates :username, :email, presence: true
+
+  before_validation ->(user) { user.username = user.email if user.username.blank? }
 end
 ```
 
-Lastly, you can define your own custom callback object, which we will cover later in more detail [below](#callback-classes).
+Lastly, you can define **a custom callback object**, as shown below. We will
+cover these [later in more detail](#callback-objects).
 
 ```ruby
 class User < ApplicationRecord
-  before_create MaybeAddName
+  validates :username, :email, presence: true
+
+  before_validation AddUsername
 end
 
-class MaybeAddName
-  def self.before_create(record)
-    if record.name.blank?
-      record.name = record.login.capitalize
+class AddUsername
+  def self.before_validation(record)
+    if record.username.blank?
+      record.username = record.email
     end
   end
 end
 ```
 
-Callbacks can also be registered to only fire on certain life cycle events, this allows complete control over when and in what context your callbacks are triggered.
+### Registering Callbacks to Fire on Lifecycle Events
```

Comment:
What do you think about explaining how to execute save and validate operations with custom contexts? Maybe this should also be included in the validation guides. For example with the `save`, `valid?` methods, etc.

Inspired by [this DHH's gist](https://gist.github.com/dhh/9672827)

Reply:
I took a pass at this [here](https://github.com/rails/rails/pull/51654/commits/802c601c7b3f531f5286bd0a9620407c3c00594f)

Reply:
Great! I would also add the example with `save`, like: `save(context: :confirmation)`, or whatever.

Do you think it is necessary to add that to the `save` method you can also pass the `touch: false` option? And that would not touch the element, therefore the callbacks would not be executed either. I'm not so sure if that escapes a little of the guide, because in fact when you do that, you are not touching the element, so you are not skipping the callback actually.

Reply:
> I would also add the example with save, like: save(context: :confirmation), or whatever.

Sounds good! I reused an example to show how you can run validations with a custom context as well as for a methis like `save`.

> Do you think it is necessary to add that to the save method you can also pass the touch: false option? 

Regarding the `touch: false option`, while it's true that it prevents the timestamp update and thus skips the associated callbacks, it's not directly related to validation contexts. Including it in the same discussion might be a bit tangential to the main topic of validation contexts, hence I decided to omit it.